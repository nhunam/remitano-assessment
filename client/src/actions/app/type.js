import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  CLEAR_MESSAGE: undefined
});
