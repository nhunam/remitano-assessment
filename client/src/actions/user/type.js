import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  USER_CREATE: undefined,
  USER_LOGIN: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined
});
