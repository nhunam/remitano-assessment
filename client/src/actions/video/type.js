import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  VIDEO_FETCH: undefined,
  VIDEO_FETCH_SUCCESS: undefined,
  VIDEO_FETCH_FAILURE: undefined,
  VIDEO_SHARING: undefined,
  VIDEO_SHARING_SUCCESS: undefined,
  VIDEO_SHARING_FAILURE: undefined
});
