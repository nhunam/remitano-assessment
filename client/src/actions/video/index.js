import { ActionTypes } from "./type";
import { createActions } from "redux-actions";

export const { videoFetch: fetch, videoSharing: shareVideo } = createActions({
  [ActionTypes.VIDEO_FETCH]: param => ({ ...param }),
  [ActionTypes.VIDEO_SHARING]: param => ({ ...param })
});
