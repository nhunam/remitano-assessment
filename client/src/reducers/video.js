import { ActionTypes } from "actions/video/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const videoState = {
  data: [],
  page: {
    total_pages: 1,
    has_next: false,
    has_previous: false,
    current_page: 0,
    total_elements: 0
  }
};

export default {
  videos: handleActions(
    {
      [ActionTypes.VIDEO_FETCH]: (state, { payload }) =>
        immutable(state, {
          data: { $set: payload.page_index === 0 ? [] : state.data },
          page: {
            $set:
              payload.page_index === 0
                ? {
                    total_pages: 1,
                    has_next: false,
                    has_previous: false,
                    current_page: 0,
                    total_elements: 0
                  }
                : state.page
          }
        }),
      [ActionTypes.VIDEO_FETCH_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          data: { $set: [...state.data, ...payload.data] },
          page: { $set: payload.page }
        }),
      [ActionTypes.VIDEO_FETCH_FAILURE]: (state, { payload }) =>
        immutable(state, {
          status: { $set: AppStatus.READY }
        })
    },
    videoState
  )
};
