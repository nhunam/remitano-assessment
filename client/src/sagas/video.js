import { all, call, put, takeLatest } from "redux-saga/effects";

import { ActionTypes } from "actions/video/type";
import { AppConfig } from "configurations";
import { parseYoutubeId } from "utils/helpers";
import { request } from "utils/client";

/**
 * Fetch Video
 */
export function* fetch(param) {
  try {
    const payload = {
      method: "GET"
    };
    const paramPayload = param.payload;
    const url = `/videos?page_index=${paramPayload.page_index}`;
    const response = yield call(request, url, payload);

    yield put({
      type: ActionTypes.VIDEO_FETCH_SUCCESS,
      payload: response
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.VIDEO_FETCH_FAILURE,
      payload: err
    });
  }
}

/**
 * Fetch Video
 */
export function* share(param) {
  try {
    const { payload } = param;
    const { videoUrl, userToken } = payload;
    const { youtubeAPIKey, youtubeAPIURL } = AppConfig;
    const requestPayload = {
      method: "GET"
    };
    const videoId = parseYoutubeId(videoUrl);
    // const origin = encodeURI(window.location.href);
    const url = `${youtubeAPIURL}?part=snippet&id=${videoId}&key=${youtubeAPIKey}`;
    const response = yield call(request, url, requestPayload);
    const { items } = response;
    const { snippet } = items[0];
    const { title, description } = snippet;

    const sharingPayload = {
      title: title,
      description: description,
      url: videoUrl
    };
    const headers = { Authorization: `Bearer ${userToken}` };
    const requestSharingPayload = {
      method: "POST",
      payload: sharingPayload,
      headers: headers
    };
    yield call(request, "/videos", requestSharingPayload);

    yield put({
      type: ActionTypes.VIDEO_SHARING_SUCCESS
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.VIDEO_SHARING_FAILURE,
      payload: err
    });
  }
}

/**
 * Video Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.VIDEO_FETCH, fetch)]);
  yield all([takeLatest(ActionTypes.VIDEO_SHARING, share)]);
}
