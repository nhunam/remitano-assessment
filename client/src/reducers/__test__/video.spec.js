import { ActionTypes } from "actions/video/type";
import reducer from "reducers";

describe("Video Reducers", () => {
  it("should return the initial state", () => {
    expect(reducer.videos(undefined, {})).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.VIDEO_FETCH}`, () => {
    expect(
      reducer.videos(undefined, {
        type: ActionTypes.VIDEO_FETCH,
        payload: { page_index: 0 }
      })
    ).toMatchSnapshot();

    expect(
      reducer.videos(undefined, {
        type: ActionTypes.VIDEO_FETCH,
        payload: { page_index: 1 }
      })
    ).toMatchSnapshot();
  });
  it(`should handle ${ActionTypes.VIDEO_FETCH_SUCCESS}`, () => {
    const actions = {
      type: ActionTypes.VIDEO_FETCH_SUCCESS,
      payload: {
        data: [],
        page: {
          total_pages: 1,
          has_next: false,
          has_previous: false,
          current_page: 0,
          total_elements: 0
        }
      }
    };
    expect(reducer.videos(undefined, actions)).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.VIDEO_FETCH_FAILURE}`, () => {
    expect(
      reducer.videos(undefined, { type: ActionTypes.VIDEO_FETCH_FAILURE })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.VIDEO_SHARING}`, () => {
    expect(
      reducer.videos(undefined, {
        type: ActionTypes.VIDEO_SHARING
      })
    ).toMatchSnapshot();
  });
  it(`should handle ${ActionTypes.VIDEO_SHARING_SUCCESS}`, () => {
    const actions = {
      type: ActionTypes.VIDEO_SHARING_SUCCESS
    };
    expect(reducer.videos(undefined, actions)).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.VIDEO_SHARING_FAILURE}`, () => {
    expect(
      reducer.videos(undefined, { type: ActionTypes.VIDEO_SHARING_FAILURE })
    ).toMatchSnapshot();
  });
});
