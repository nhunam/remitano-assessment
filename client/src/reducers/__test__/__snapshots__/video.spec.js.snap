// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Reducers should handle VIDEO_FETCH 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;

exports[`Video Reducers should handle VIDEO_FETCH 2`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;

exports[`Video Reducers should handle VIDEO_FETCH_FAILURE 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
  "status": undefined,
}
`;

exports[`Video Reducers should handle VIDEO_FETCH_SUCCESS 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;

exports[`Video Reducers should handle VIDEO_SHARING 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;

exports[`Video Reducers should handle VIDEO_SHARING_FAILURE 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;

exports[`Video Reducers should handle VIDEO_SHARING_SUCCESS 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;

exports[`Video Reducers should return the initial state 1`] = `
Object {
  "data": Array [],
  "page": Object {
    "current_page": 0,
    "has_next": false,
    "has_previous": false,
    "total_elements": 0,
    "total_pages": 1,
  },
}
`;
